CMS, OSD and MSP_DISPLAYPORT (grafting from pre-BF3.1)

jflyper opened the pull request:
This is a combined PR for CMS, OSD and MSP_DISPLAYPORT.

### Thing added
- CMS (Configuration Menu System) that can run on multiple devices, including
- Integrated OSD based on MAX7456 (a new driver provided with this PR; OMNIBUS target atm)
- External OSD (MWOSD atm) using MSP_DISPLAYPORT, and
- OLED DASHBOARD (renamed from DISPLAY).
- Can cycle through multiple display devices if they are active at the same time (e.g., External OSD and DASHBOARD, or Integrated OSD and DASHBOARD), by entering menu invocation command while in menu.

### Target configuration
CMS is currently defined for all 128K+ targets, and External OSD and Integrated-OSD are mutually exclusive. This means that CMS enabled targets without an Integrated OSD will automatically be ready for External OSD, and targets with Integrated OSD will automatically disable External OSD.
This configuration scheme was selected to save a feature bit (there are  only couples left), and  is open for discussion.

### CMS invocation
To enter the CMS, enter Throttle MID + Yaw LEFT + Pitch FULL.

### Using External-OSD (MSP_DISPLAYPORT)
To use the MSP_DISPLAYPORT device, corresponding MWOSD must be installed in the MinimOSD compatible device:
[https://github.com/ShikOfTheRa/scarab-osd/tree/master](https://github.com/ShikOfTheRa/scarab-osd/tree/master)
Please be careful not to enter the CMS while MWOSD menu is active (MWOSD's menu invocation is Yaw RIGHT.)

### Dual CMS chaos
When the FC-side CMS is enabled, you are going to have two CMS in your system; one in the FC and one in MWOSD. CMS invocation stick commands should be different, but you must be aware that these two CMS may not have the same navigational model and command system. You have to deal with this duality.

### Menu content
Menu content in the initial PR is reductive port of the BF version, and will be expanded to cover standard as well as iNav specific configurable parameters.

### Configurator support for OSD
This is a separate to do on the configurator. It may include the configuration of the CMS itself depending on the discussion mentioned above.

ezshinoda committed:
Brute force grafting

Attempt to brutally graft BF3.1’s OSD, CMS, DISPLAYPORT, DASHBOARD,
MSP_DISPLAYPORT onto INAV-1.4.

With this first cut, integrated OSD (based on MAX7456) and CMS over the
OSD is working.

martinbudden committed:
MSP catchup with betaflight

ezshinoda committed:
Revive GPS numsat and speed, deleted logo

It’s cosmetically nice if we have a bitmap logo, but my guess is that
it will going to be smaller than that of BF, since there will more
symbols required for nav related displays.

ezshinoda committed:
display renamed to dashboard (per BF) and operational under CMS

ezshinoda committed:
Makefile, target/common.h to be coherent with BF

ezshinoda committed:
Rebased on to up to date master

martinbudden commented on src/main/target/SPRACINGF3/target.h:
```diff
@@ -126,3 +126,5 @@
 #define USABLE_TIMER_CHANNEL_COUNT 17
 #define USED_TIMERS             ( TIM_N(1) | TIM_N(2) | TIM_N(3) | TIM_N(4) | TIM_N(15) | TIM_N(16) | TIM_N(17) )
 
+#define USE_MSP_DISPLAYPORT
```
Not required, since this is in `common.h`

jflyper replied:
ack

jflyper replied:
Deleted. Will be updated on next sync.

martinbudden commented on src/main/io/dashboard.c~HEAD:
```diff
@@ -0,0 +1,752 @@
+/*
```
Not sure how this file got into the PR, it needs to be removed.

jflyper replied:
Wow!?

jflyper replied:
Deleted dashboard.[ch]~HEAD.

ezshinoda committed:
Added RATE & EXPO menu, fixed dashboard to display page title in post-grab case

ezshinoda committed:
Remaining PID menu is now in.

ezshinoda committed:
Roll, Pitch and Yaw rates are now displayed in deg/sec

ezshinoda committed:
Code tidy

DzikuVx commented on src/main/cms/cms_menu_imu.c:
```diff
@@ -316,11 +316,11 @@ static OSD_Entry cmsx_menuFilterGlobalEntries[] =
-    { "GYRO NF2",   OME_UINT16, NULL, &(OSD_UINT16_t) { &masterConfig.gyro_soft_notch_hz_2,     0, 500, 1 }, 0 },
-    { "GYRO NF2C",  OME_UINT16, NULL, &(OSD_UINT16_t) { &masterConfig.gyro_soft_notch_cutoff_2, 0, 500, 1 }, 0 },
+    { "GYRO LPF",   OME_UINT8,  NULL, &(OSD_UINT8_t){ &masterConfig.gyro_soft_lpf_hz,          0, 255, 1 }, 0 },
+    { "GYRO NF1",   OME_UINT16, NULL, &(OSD_UINT16_t){ &masterConfig.gyro_soft_notch_hz_1,     0, 500, 1 }, 0 },
```
We do not have notch filters yet, so this should be commented out

jflyper replied:
Yes. It will be gone on the next push.

The the notch section is in the big #ifdef NOT_YET - #endif now. It was invalidated at the start when I was an absolute noob on iNav's imu. Now I'm certain that it doesn't exist :)

DzikuVx commented on src/main/cms/cms_menu_imu.c:
```diff
@@ -316,11 +316,11 @@ static OSD_Entry cmsx_menuFilterGlobalEntries[] =
-    { "GYRO NF1C",  OME_UINT16, NULL, &(OSD_UINT16_t) { &masterConfig.gyro_soft_notch_cutoff_1, 0, 500, 1 }, 0 },
-    { "GYRO NF2",   OME_UINT16, NULL, &(OSD_UINT16_t) { &masterConfig.gyro_soft_notch_hz_2,     0, 500, 1 }, 0 },
-    { "GYRO NF2C",  OME_UINT16, NULL, &(OSD_UINT16_t) { &masterConfig.gyro_soft_notch_cutoff_2, 0, 500, 1 }, 0 },
+    { "GYRO LPF",   OME_UINT8,  NULL, &(OSD_UINT8_t){ &masterConfig.gyro_soft_lpf_hz,          0, 255, 1 }, 0 },
```
this is 0-5 range

ezshinoda committed:
Menu restructuring, adding & deleting entries, etc., etc., ...

ezshinoda committed:
IMU menu entry ordering changed

It’s now
PID (profile dependent)
PID ALTMAG (profile dependent)
PID GPSNAV (profile dependent)
FILT PP (profile dependent)
RATE (rate profile dependent)
GYRO (global)

ezshinoda committed:
Auto repeat for left and right keys

Key repeat effect is implemented in two phases.
First phldase is to decrease rcDelayMs reciprocal to hold time.
When rcDelayMs reached a certain limit (scheduling interval),
repeat rate will not raise anymore, so we call key handler
multiple times (repeatCount).

XXX Caveat: Most constants are adjusted pragmatically.
XXX Rewrite this someday, so it uses actual hold time instead of
holdCount, which depends on the scheduling interval.

ezshinoda committed:
Merge remote-tracking branch 'inav/development' into inavdev-cms-osd-displayport

ezshinoda committed:
Activate MAX7456 and OSD on OMNIBUSF4

ezshinoda committed:
2 more chars for right column (value field)
